File: python/815_BusRoutes.py
from unittest import TestCase
# https://leetcode.com/problems/bus-routes/
import collections


class BusRoutes(TestCase):
    def numBusesToDestination(self, routes: 'List[List[int]]', S: 'int', T: 'int') -> 'int':
        if S == T: return 0
        n, done, que = len(routes), set(), collections.deque()
        edges = [[] for _ in range(n)]
        for i in range(len(routes)):
            routes[i] = set(routes[i])
            if S in routes[i]:
                que.appendleft((1, i))
                done.add(i)
        for i in range(n):
            for j in range(i+1, n):
                if any(s in routes[j] for s in routes[i]):
                    edges[i].append(j)
                    edges[j].append(i)
        while que:
            cnt, i = que.pop()
            if T in routes[i]:
                return cnt
            for j in edges[i]:
                if j not in done:
                    done.add(j)
                    que.appendleft((cnt+1, j))
        return -1

    def test1(self):
        self.assertEqual(2, self.numBusesToDestination([[1, 2, 7], [3, 6, 7]], 1, 6))

File: kotlin/src/main/kotlin/815.BusRoutes.kt
import org.junit.Assert.assertEquals
import org.junit.Test
import java.util.*

//https://leetcode.com/problems/bus-routes/


class BusRoutes{
  fun numBusesToDestination(routes: Array<IntArray>, S: Int, T: Int): Int {
    if(S == T) return 0
    val n = routes.size
    val rt = routes.map{ it.toSet() }
    val edges = (0 until n).map{ mutableListOf<Int>() }
    val que = LinkedList<Pair<Int, Int>>()
    for(i in 0 until n){
      if(S in rt[i]) que.push(1 to i)
      for(j in i+1 until n)
        if(routes[i].any{it in rt[j]}) {
          edges[i].add(j)
          edges[j].add(i)
        }
    }

    val done = que.map{it.second}.toMutableSet()
    while(que.isNotEmpty()){
      val (cnt, i) = que.pollLast()
      if(T in rt[i]) return cnt
      for(j in edges[i])
        if(j !in done) {
          done.add(j)
          que.push(cnt + 1 to j)
        }
    }

    return -1
  }

  @Test
  fun test1(){
    assertEquals(2, numBusesToDestination(arrayOf(intArrayOf(1, 2, 7), intArrayOf(3, 6, 7)), 1, 6))
  }
}